saw-central/src/SAWCentral/Crucible/MIR/Override.hs

```diff
@@ -1029,53 +1029,100 @@ learnSetupCondition opts sc cc spec prepost cond =
 -- | Match the value of a function argument with a symbolic 'SetupValue'.
 matchArg ::
   forall w.
-  Options          {- ^ saw script print out opts -} ->
-  SharedContext      {- ^ context for constructing SAW terms    -} ->
-  MIRCrucibleContext    {- ^ context for interacting with Crucible -} ->
-  CrucibleMethodSpecIR {- ^ specification for current function override  -} ->
-  MS.PrePost                                                          ->
+  Options              {- ^ saw script print out opts -} ->
+  SharedContext        {- ^ context for constructing SAW terms -} ->
+  MIRCrucibleContext   {- ^ context for interacting with Crucible -} ->
+  CrucibleMethodSpecIR {- ^ specification for current function override -} ->
+  MS.PrePost ->
   MS.ConditionMetadata ->
-  MIRVal             {- ^ concrete simulation value             -} ->
-  Mir.Ty             {- ^ expected memory type                  -} ->
-  SetupValue         {- ^ expected specification value          -} ->
+  MIRVal               {- ^ concrete simulation value -} ->
+  Mir.Ty               {- ^ expected memory type -} ->
+  SetupValue           {- ^ expected specification value -} ->
+  OverrideMatcher MIR w ()
+matchArg opts sc cc cs prepost md actual expectedTy expected =
+  matchArg' opts sc cc cs prepost md actual expectedTy expected False
+
+-- | The implementation of matchArg, with an additional argument to keep track
+-- of whether the given 'SetupValue' is contained inside a 'SetupCast'. The
+-- 'SetupVar' case uses different types depending on whether it is in a
+-- 'SetupCast' or not, but the 'SetupCast' case can't just recursively call
+-- 'matchArg' with different types, since we don't know what the types should be
+-- until we get to the 'SetupVar'. So instead we just pass along that fact that
+-- we are recursing inside a 'SetupCast'.
+matchArg' ::
```

Comment:
Instead of defining an entirely separate, top-level `matchArg'` function, I think we should just define `matchArg` like so:

```hs
matchArg opts sc cc cs prepost md = go True
  where
    go :: Bool -> MIRVal -> Mir.Ty -> SetupValue -> OverrideMatcher MIR w ()
    go inCast actual expectedTy expected = ...
```

Where `go` is the recursive function that does what `matchArg'` does currently. This keeps the logic for this function contained within a single `matchArg` function, and it makes it more obvious that the only values that can change on a recursive call are `inCast`, `actual`, `expectedTy`, and `expected`.